DESIGN.md

```diff
@@ -0,0 +1,79 @@
+# RPC
+This project implements an RPC server in Go using gRPC as a transport.
+
+# Design
+What follows is a design document outlining the technology choices, architecture, and implementation of the RPC client and server. 
+
+## Authentication and Authorization
+The RPC service includes mutual TLS authentication. The client and server will both have their own private keys as well as certificates signed by the same CA. The minimum TLS version will be set to TLS1.2. In order to protect against downgrade attacks in TLS1.2, an allowlist will be used for ciphers. Conveniently, the Go TLS package will return a list of all cipher suites it implements excluding suites with known security issues (tls.CipherSuites()). In order to restrict the client to using these only, PreferServerCipherSuites will be set to true in the server TLS configuration. If this were a production application, an endpoint would be included in the API to accept a certificate revocation list, allowing the administrator to 'ban' certain clients. 
+
+For Authorization, a simple form of role based access control will be implemented: Each role will have its own CA. All CAs will be trusted by the server. Upon handling a remote procedure call, the server will first check the client certificate's certificate authority and compare it to its internal mapping of certificate authority to privilege level. The following roles will be implemented:
```

Comment:
can you in a couple of sentences expand if you would use a form of this in a production environment?
if yes why did you choose it, if no what's the alternative.

Reply:
I'm also interested in discussions of tradeoffs for the current system -- not that it needs to be reformed -- I'm only curious about the implications and alternatives.

Reply:
I would not use this in a production environment. In production I would prefer to use a hierarchical CA instead of the disjoint ones.

A Hierarchical CA would be useful here in that it would allow for multiple client CAs signing certificates and the server would only need the root certificate. Another advantage of this approach is that compromised client CAs could be taken offline without having to update the certificates loaded by the server beyond the updated revocation list from the root CA.

Reply:
I am sorry I didn't make myself clear, I was mostly asking about using CAs for authz. Does that change your answer above?

Reply:
Thanks for the clarification. My answer is still largely the same. I am familiar with using Oauth, Radius, LDAP, and other identity based solutions which I would prefer to use in a production application. These solutions all involve leveraging external services, so I do not consider them in scope for this project.

Using the Certificates like this allows me to get away with not storing user credentials or secrets. Storing passwords in general can be tricky to do safely. Since I am already issuing client certificates, and I can differentiate between client certificates trusted by multiple CAs, I can use this to authorize different client connections.

Reply:
So to helicopter in on a couple things. 

First, I'm fine seeing what the solution looks like with the separate CA's for different roles as separate roots of trusts. While this may be straying into a customized security model, I can see it as a plausible deployment model in fairly narrow use cases. However, I would mention that I likely wouldn't accept this model at higher levels.

There are what I think of as 3 main / common approaches to this type of problem that may not have been considered, which I think is what the team is trying to get at.
1. Embed role information in the certificates. This is likely the most common, and is the main deployment mode for machine to machine trust in kubernetes for example. User information gets embedded in CN, and group information in the O or OU fields which are repeatable. Because the contents of the cert are also verified by the trust building in TLS, these fields can be a secured indication of user and group or even arbitrary data. Outside of internal systems I think the kubernetes docs provide some insight into this model: https://kubernetes.io/docs/setup/best-practices/certificates/
2. Embed only a username in the certificate CN, and then use a user database to retrieve role / permission / group / revocation information. This sort of bypasses the distributed trust model of TLS, but can be appropriate for internal tooling that don't need a distributed trust from the CA to the implementation. As applicable to this project, it would just mean pulling the username from the cert and applying some hard coded rules, user1 = ro, user2=admin, etc, etc.
3. Generate only self signed certificates, and add just the self signed certs to the trust store. You don't actually need the path building and signing, and can just authenticate/authorize on a preloaded cert from a client if it's static in most cases. And similar, use a hardcoded permissions based on the self signed cert.


To sort of poke at the preference for the listed identity solutions... there are a few reasons we think mTLS is the right answer for service to server and most of user to service authentication and authorization, and as such are promoters of this model over direct integration of services with Oauth, Radius, and LDAP. We still connect to identity providers, but we would connect the CA to issue certificates... or use more than one source for authentication. 

The reason we then like certificates is they have a few properties most identity protocols don't offer. Without making this a huge manifesto, the main property advantage that we like is that mTLS uses proofs to authenticate. This means a compromised endpoint cannot impersonate a user by having the user connect. Many of the more traditional central identity protocols don't have this property, logging into a server you don't control is inherently dangerous, as you're trusting the admin not to clone your credentials, or the server not to be running malware or otherwise expose a set of credentials. Many token based systems suffer in similar ways, where a token is a short lived but reusable secret that other systems may be configured to accept.

This is one of the reasons why we like mTLS, and our product suite is based around this concept, and we're doing what we can to push wider adoption.

Reply:
Thanks @knisbet,

I considered going the route of #2, but preferred a model where I wouldn't have to track usernames. Being able to use the CA certificate to determine user role without having to keep track of users in a database or configuration file seemed like a logical iteration on that design.

I also considered #3, but it seemed like a better technical decision to leverage CAs because then certificates can be generated without modifying the server's trust store.

That being said, I like option #1 and would like to modify my design to use it. It seems simpler than my multiple CA design and equally as effective. I will modify my design document for that change.